### PR TITLE
camera: camera refinement

### LINKED ISSUE #3763: System camera is mirrored
With the system camera, the image is mirrored. This is shown in the preview and game. Using Ultra Moon as an example. Tested with 80bfd87.

citra-qt_2018-05-19_20-01-21

[Log](https://pastebin.com/rsGRbw8r)

### LINKED ISSUE #3797: AR Games - Green Screen with System Camera (qt)
Reporting what was posted in the [forum](https://community.citra-emu.org/t/system-camera-qt-shows-green-screen-on-ar-games/23507). This used to work in previous Canary builds. The last one I could find before this happened was Canary-450 (b40c597) though the image is flipped upside down. Note that I had to copy the mediaservice folder with a DLL file from a newer build for this to work since it was missing.

Tested with Nightly-705 (72f9142)
[Log](https://pastebin.com/nNfPRM6a)

citra-qt_2018-05-31_16-01-11
citra-qt_2018-05-31_16-28-18

### PR DESCRIPTION
**Pull request by zhaowenlan1779**
Pretty self-explanatory.
closes #3763, closes #3797

This contains:
- [X] camera flip config
- [X] camera interface refactor
- [X] camera: migrate logging macros
- [X] camera: fix QCamera multiple load problem
- [X] move StillImageCameraFactory::GetFilePath to Ui thread (you can't actually call any UI-related things safely in Qt in non-UI thread)
- [X] fix `invalid viewfinder settings` problem
- [x] fix StillImageCamera multiple prompt problem

With this PR, camera should become a _nearly_ completed feature, (except the issues mentioned in #3739)

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/3764)

### CODE REVIEW

File: src/citra_qt/configuration/configure_camera.cpp [hunk 1]
```diff
@@ -27,14 +27,7 @@ ConfigureCamera::ConfigureCamera(QWidget* parent)
     // Load settings
     camera_name = Settings::values.camera_name;
     camera_config = Settings::values.camera_config;
-    for (auto&& item : camera_name) {
-        if (item == "opencv") {
-            QMessageBox::critical(this, tr("Error"),
```
Comment (baka0815) [inline src/citra_qt/configuration/configure_camera.cpp]: Why did you remove this?

File: src/citra_qt/camera/still_image_camera.cpp [hunk 2]
```diff
@@ -26,8 +31,8 @@ void StillImageCamera::SetResolution(const Service::CAM::Resolution& resolution)
 
 void StillImageCamera::SetFlip(Service::CAM::Flip flip) {
     using namespace Service::CAM;
-    flip_horizontal = (flip == Flip::Horizontal) || (flip == Flip::Reverse);
-    flip_vertical = (flip == Flip::Vertical) || (flip == Flip::Reverse);
+    flip_horizontal = basic_flip_horizontal ^ (flip == Flip::Horizontal || flip == Flip::Reverse);
+    flip_vertical = basic_flip_vertical ^ (flip == Flip::Vertical || flip == Flip::Reverse);
```
Comment (baka0815) [inline src/citra_qt/camera/still_image_camera.cpp]: Can't ypu create a method for this? You're doing it multiple times, so changing the logig (like you're doing now) needs to be applied to all places.

File: src/citra_qt/configuration/configure_camera.ui [hunk 3]
```diff
@@ -1,257 +1,329 @@
 <?xml version="1.0" encoding="UTF-8"?>
-
 <ui version="4.0">
-  <class>ConfigureCamera</class>
-  <widget class="QWidget" name="ConfigureCamera">
-    <layout class="QVBoxLayout" name="verticalLayout">
+ <class>ConfigureCamera</class>
+ <widget class="QWidget" name="ConfigureCamera">
```
Comment (baka0815) [inline src/citra_qt/configuration/configure_camera.ui]: Whitespace changes?
Could you check and change? Reviewing this file like this is really hard.

File: src/citra_qt/configuration/configure_camera.cpp (hunk 1 above)
Comment (zhaowenlan1779) [reply src/citra_qt/configuration/configure_camera.cpp]: This was intended to be temporary. It was just to inform users from old canary versions that OpenCV has been removed. Now I feel it safe to remove that.

File: src/citra_qt/camera/still_image_camera.cpp (hunk 2 above)
Comment (zhaowenlan1779) [reply src/citra_qt/camera/still_image_camera.cpp]: OK I will.

File: src/citra_qt/configuration/configure_camera.ui (hunk 3 above)
Comment (zhaowenlan1779) [reply src/citra_qt/configuration/configure_camera.ui]: Do you know you can ```Hide Whitespace Changes``` in ```Diff settings```?
Anyway, I'm going to remove whitespace changes

**Comment by zhaowenlan1779** [comment]
The CI failure is not my bad..

File: src/citra_qt/configuration/configure_camera.cpp (hunk 1 above)
Comment (baka0815) [reply src/citra_qt/configuration/configure_camera.cpp]: Ok, I was just irritated because I just translated that text. :-)

File: src/citra_qt/configuration/configure_camera.ui (hunk 3 above)
Comment (baka0815) [reply src/citra_qt/configuration/configure_camera.ui]: Ni, I didn't know this (and it's hidden from the mobile version).
Looks fine then.

File: src/citra_qt/camera/qt_camera_base.cpp [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+}
+
+void QtCameraInterface::SetResolution(const Service::CAM::Resolution& resolution) {
+    width = resolution.width;
+    height = resolution.height;
+}
+
+void QtCameraInterface::SetFlip(Service::CAM::Flip flip) {
+    using namespace Service::CAM;
+    flip_horizontal = basic_flip_horizontal ^ (flip == Flip::Horizontal || flip == Flip::Reverse);
```
Comment (wwylele) [inline src/citra_qt/camera/qt_camera_base.cpp]: The logical XOR in C++ is `!=` operator. I am not sure if others might prefer bitwise XOR `^` for code readability, though...

File: src/citra_qt/camera/qt_multimedia_camera.cpp [hunk 5]
```diff
@@ -104,27 +97,9 @@ void QtMultimediaCamera::SetFrameRate(Service::CAM::FrameRate frame_rate) {
...
-        NGLOG_ERROR(Service_CAM, "Unimplemented effect {}", static_cast<int>(effect));
-    }
-}
-
-std::vector<u16> QtMultimediaCamera::ReceiveFrame() {
+const QImage& QtMultimediaCamera::QtReceiveFrame() {
     QMutexLocker locker(&handler->camera_surface.mutex);
-    return CameraUtil::ProcessImage(handler->camera_surface.current_frame, width, height,
-                                    output_rgb, flip_horizontal, flip_vertical);
+    return handler->camera_surface.current_frame;
```
Comment (wwylele) [inline src/citra_qt/camera/qt_multimedia_camera.cpp]: `camera_surface.current_frame` is returned as a reference, but its associate mutex is unlocked outside this function. Therefore the caller (`QtCameraInterface::ReceiveFrame`) would cause race condition when using this reference. 

To keep your interface structure, you need to change the return type of `QtReceiveFrame` to value-return, so that it can make a copy of the image when unlocking the mutex. If you want to avoid copying, my suggestion to resolve this is to make a member `QtCameraInterface::ProcessFrame(const QImage& image){ CameraUtil::ProcessImage(image, width, height, output_rgb, flip_horizontal, flip_vertical);}`and let subclasses call this in `ReceiveFrame()`, instead of virtual dispatching the function in the other way around.

File: src/citra_qt/camera/qt_camera_base.cpp (hunk 4 above)
Comment (zhaowenlan1779) [reply src/citra_qt/camera/qt_camera_base.cpp]: I would prefer ```^```, but if you think ```!=``` is better I will change that

File: src/citra_qt/camera/qt_multimedia_camera.cpp (hunk 5 above)
Comment (zhaowenlan1779) [reply src/citra_qt/camera/qt_multimedia_camera.cpp]: Well, I thought about this but don't think it's a major problem..
OK. I would say I know almost nothing about multithread stuff.

Comment (zhaowenlan1779) [reply src/citra_qt/camera/qt_multimedia_camera.cpp]: BTW, 
> QImage objects can be passed around by value since the QImage class uses implicit data sharing. QImage objects can also be streamed and compared.

(qt doc)

**Comment by zhaowenlan1779** [comment]
@wwylele Addressed your comment

File: src/core/settings.h [hunk 6]
```diff
@@ -140,6 +140,7 @@ struct Values {
     // Camera
     std::array<std::string, Service::CAM::NumCameras> camera_name;
     std::array<std::string, Service::CAM::NumCameras> camera_config;
+    std::array<int, Service::CAM::NumCameras> camera_flip;
```
Comment (wwylele) [inline src/core/settings.h]: This should probably be moved to `UISettings` in Qt as it is not for communication with the core and is a Qt-specific thing.

Comment (zhaowenlan1779) [reply src/core/settings.h]: @wwylele But doesn't UISettings mean "settings for the UI"? Is it really suitable?

Comment (zhaowenlan1779) [reply src/core/settings.h]: @wwylele But doesn't UISettings mean "settings for the UI"? Is it really suitable?

Comment (wwylele) [reply src/core/settings.h]: @zhaowenlan1779 we need to rename it in the future then.

Comment (baka0815) [reply src/core/settings.h]: Currently it's Qt specific, but won't it be useful in the SDL2 frontend in the future?

Comment (wwylele) [reply src/core/settings.h]: @baka0815 then there should be a common layer shared by the two frontend, without penetrating the core

**Comment by zhaowenlan1779** [comment]
@wwylele What's up?

File: src/core/settings.h (hunk 6 above)
Comment (zhaowenlan1779) [reply src/core/settings.h]: But camera flip is actually retrieved in the core.

Comment (wwylele) [reply src/core/settings.h]: @zhaowenlan1779 oh I see... this is fine to me then

**Comment by baka0815** [comment]
@zhaowenlan1779 he edited his comment. It was about the above nit previously.

**Comment by zhaowenlan1779** [comment]
Please test and tell me if there is any issues with my new commit e102392

**Comment by zhaowenlan1779** [comment]
Rebased to get commit history clear

File: src/core/frontend/camera/factory.h [hunk 7]
```diff
@@ -22,7 +22,7 @@ class CameraFactory {
      * @returns a unique_ptr to the created camera object.
      */
     virtual std::unique_ptr<CameraInterface> Create(const std::string& config,
-                                                    const Service::CAM::Flip& flip) const = 0;
+                                                    const Service::CAM::Flip& flip) = 0;
```
Comment (zhaowenlan1779) [inline src/core/frontend/camera/factory.h]: Have to remove `const` to get `this` as normal pointer to pass to `QMetaObject::invokeMethod`

File: src/citra_qt/camera/still_image_camera.cpp [hunk 8]
```diff
@@ -24,7 +25,7 @@ bool StillImageCamera::IsPreviewAvailable() {
     return !image.isNull();
 }
 
-const std::string StillImageCameraFactory::GetFilePath() {
+const std::string StillImageCameraFactory::GetFilePath() const {
```
Comment (zhaowenlan1779) [inline src/citra_qt/camera/still_image_camera.cpp]: I'm trying to make things `const` where possible

**Comment by Leo626** [comment]
This fixes AR Games. Also I found that Pokemon ORAS is another game that has the green screen issue but for some reason is not fixed with this PR. It used to work in previous Canary builds.

[Log](https://pastebin.com/3Jcn5RVJ)

citra-qt_2018-06-03_11-01-32
citra-qt_2018-06-03_14-26-39

**Comment by Leo626** [comment]
ORAS is fixed now.
citra-qt_2018-06-07_07-48-51
